src/modules/Elsa.Workflows.Core/Contexts/WorkflowExecutionContext.Cancel.cs

```diff
@@ -0,0 +1,43 @@
+using Elsa.Extensions;
+
+namespace Elsa.Workflows;
+
+/// <summary>
+/// Provides context to the currently executing workflow.
+/// </summary>
+public partial class WorkflowExecutionContext
+{
+    private ICollection<CancellationTokenSource> _cancellationTokenSources = new List<CancellationTokenSource>();
+    private ICollection<CancellationTokenRegistration> _cancellationRegistrations = new List<CancellationTokenRegistration>();
+
+    /// <summary>
+    /// Cancels the workflow and all it's children.
+    /// </summary>
+    public void Cancel()
+    {
+        foreach (var source in _cancellationTokenSources)
+            source.Cancel();
+        
+        _cancellationTokenSources.Clear();
+    }
+    
+    private void CancelWorkflow()
+    {
+        Bookmarks.Clear();
+        _completionCallbackEntries.Clear();
+
+        if (Status != WorkflowStatus.Running && SubStatus != WorkflowSubStatus.Faulted)
+            return;
+
+        AddExecutionLogEntry("Workflow cancelled");
+        // How to guarantee persistence of bookmarks??
```

Comment:
I'm wondering if we even need to persist bookmarks at this point, since we're cancelling the workflow anyway. Do you have a scenario in mind where the persistence of bookmarks is important?

Reply:
Persistence of bookmarks in the case of cancellation would mean cleaning up the database, Not terribly important, but might be good for consistency.